WebDriverAgentLib/Categories/XCUIElement+FBFind.m

```diff
@@ -83,18 +85,43 @@ - (void)descendantsWithProperty:(NSString *)property value:(NSString *)value par
   // XPath will try to match elements only class name, so requesting elements by XCUIElementTypeAny will not work. We should use '*' instead.
   xpathQuery = [xpathQuery stringByReplacingOccurrencesOfString:@"XCUIElementTypeAny" withString:@"*"];
   NSArray *matchingSnapshots = [self.lastSnapshot fb_descendantsMatchingXPathQuery:xpathQuery];
-  NSArray *allElements = [[self descendantsMatchingType:XCUIElementTypeAny] allElementsBoundByIndex];
-  NSArray *matchingElements = [self filterElements:allElements matchingSnapshots:matchingSnapshots];
+  if (0 == [matchingSnapshots count]) {
+    return @[];
+  }
+  // Prefiltering elements speeds up search by XPath a lot, because [element resolve] is the most expensive operation here
+  NSSet *matchedTypes = [XCUIElement getSnapshotsTypes:matchingSnapshots];
+  NSDictionary *elementsByType = [self splitElementsByType:matchingSnapshots byTypes:matchedTypes];
+  NSArray *matchingElements = [self filterElements:elementsByType matchingSnapshots:matchingSnapshots];
   return matchingElements;
 }
 
-- (NSArray<XCUIElement *> *)filterElements:(NSArray<XCUIElement *> *)elements matchingSnapshots:(NSArray<XCElementSnapshot *> *)snapshots
++(NSSet *)getSnapshotsTypes:(NSArray<XCElementSnapshot *> *)matchingSnapshots
```

Comment:
1) This should go to `XCElementSnapshot`. 
2) `NSSet<NSNumber *> *`
3) Write tests